src/ray/core_worker/core_worker.cc

```diff
@@ -2384,7 +2403,9 @@ void CoreWorker::HandleSubscribeForObjectEviction(
     RAY_LOG(DEBUG) << stream.str();
     send_reply_callback(Status::NotFound(stream.str()), nullptr, nullptr);
   } else {
-    object_status_publisher_->RegisterSubscription(subscriber_node_id, object_id);
+    object_status_publisher_->RegisterSubscription(
+        rpc::ChannelType::WAIT_FOR_OBJECT_EVICTION, subscriber_node_id,
+        object_id.Binary());
```

Comment:
Hmm I don't know which refactoring you are talking about, but I don't think this scenario is possible (unless you have 1~2 cpus in your machine. And although there are 1\~2 cpus, it is really unlikely because the class is thread-safe). If you are really concerned, I can just modify it because it is not a big change.

Reply:
The task execution thread runs concurrently with the asio IO context thread that this gRPC handler will be run on, so couldn't the task execution thread invoke the delete callback before the subscription is registered, therefore leaking the subscription since it will never get unregistered? I don't see any synchronization that would prevent the delete callback that's added in line 2408 being run before the subscription registration is added in line 2414.

Implicitly assuming that the delete callback (which unregisters the subscription) won't be invoked until after a later subscription registration, in the face of unsynchronized concurrent calling of the delete callback and registering of the subscription, is a fragile assumption that could be easy to miss on a refactor. Even if this is very unlikely or even impossible with the current higher-level design of the core worker, for the sake of being defensive towards future refactors, it makes a lot more sense to me to register the subscription before the delete callback is registered.